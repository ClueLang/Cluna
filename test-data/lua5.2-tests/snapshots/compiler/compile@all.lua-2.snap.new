---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
