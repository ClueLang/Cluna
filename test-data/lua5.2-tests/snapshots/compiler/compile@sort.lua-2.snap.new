---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
