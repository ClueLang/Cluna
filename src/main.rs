#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
