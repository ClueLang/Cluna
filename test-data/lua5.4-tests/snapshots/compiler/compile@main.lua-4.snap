---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.4-tests/main.lua
---
if _port {
    return
}
print("testing stand-alone interpreter");
assert(os.execute());
local arg = arg || ARG
local prog = os.tmpname()
local otherprog = os.tmpname()
local out = os.tmpname()
local progname
{
    local i = 0
    while arg[(i)] {
        i = i - 1
    }
    progname = arg[(i + 1)]
}
print("progname: " .. progname);
local prepfile = fn (s, p) {
    p = p || prog
    io.output(p);
    io.write(s);
    assert(io.close());
}
local fn getoutput() {
    io.input(out);
    local t = io.read("a")
    io.input()::close();
    assert(os.remove(out));
    return t
}
local fn checkprogout(s) {
    assert(string.sub(s, -1) == "\n");
    local t = getoutput()
    for line with string.gmatch(s, ".-\n") {
        assert(string.find(t, line, 1, true));
    }
}
local fn checkout(s) {
    local t = getoutput()
    if s != t {
        print(string.format("'%s' - '%s'\n", s, t));
    }
    assert(s == t);
    return t
}
local fn RUN(p, ...) {
    p = string.gsub(p, "lua", '"' .. progname .. '"', 1)
    local s = string.format(p, ...)
    assert(os.execute(s));
}
local fn NoRun(msg, p, ...) {
    p = string.gsub(p, "lua", '"' .. progname .. '"', 1)
    local s = string.format(p, ...)
    s = string.format("%s 2> %s", s, out)
    assert(notos.execute(s));
    assert(string.find(getoutput(), msg, 1, true));
}
RUN('lua -v');
print(string.format("(temporary program file used in these tests: %s)", prog));
prepfile("");
RUN('lua - < %s > %s', prog, out);
checkout("");
prepfile([[
  print(
1, a
)
]]);
RUN('lua - < %s > %s', prog, out);
checkout("1\tnil\n");
RUN('echo "print(10)\nprint(2)\n" | lua > %s', out);
checkout("10\n2\n");
RUN('echo "print(arg[1])" | lua - -h > %s', out);
checkout("-h\n");
prepfile("print(package.path)");
RUN('env LUA_INIT= LUA_PATH=x lua %s > %s', prog, out);
checkout("x\n");
RUN('env LUA_INIT= LUA_PATH_5_4=y LUA_PATH=x lua %s > %s', prog, out);
checkout("y\n");
prepfile("print(package.cpath)");
RUN('env LUA_INIT= LUA_CPATH=xuxu lua %s > %s', prog, out);
checkout("xuxu\n");
RUN('env LUA_INIT= LUA_CPATH_5_4=yacc LUA_CPATH=x lua %s > %s', prog, out);
checkout("yacc\n");
prepfile("print(X)");
RUN('env LUA_INIT="X=tonumber(arg[1])" lua %s 3.2 > %s', prog, out);
checkout("3.2\n");
prepfile("print(X)");
RUN('env LUA_INIT_5_4="X=10" LUA_INIT="X=3" lua %s > %s', prog, out);
checkout("10\n");
prepfile("x = x or 10; print(x); x = x + 1");
RUN('env LUA_INIT="@%s" lua %s > %s', prog, prog, out);
checkout("10\n11\n");
NoRun('LUA_INIT:1: msg', 'env LUA_INIT="error(\'msg\')" lua');
local defaultpath, defaultCpath
{
    prepfile("print(package.path, package.cpath)");
    RUN('env LUA_INIT="error(10)" LUA_PATH=xxx LUA_CPATH=xxx lua -E %s > %s', prog, out);
    local output = getoutput()
    defaultpath = string.match(output, "^(.-)\t")
    defaultCpath = string.match(output, "\t(.-)$")
    RUN('env -i lua %s > %s', prog, out);
    local out = getoutput()
    assert(defaultpath == string.match(output, "^(.-)\t"));
    assert(defaultCpath == string.match(output, "\t(.-)$"));
}
assert(notstring.find(defaultpath, "xxx") && string.find(defaultpath, "lua") && notstring.find(defaultCpath, "xxx") && string.find(defaultCpath, "lua"));
local fn convert(p) {
    prepfile("print(package.path)");
    RUN('env LUA_PATH="%s" lua %s > %s', p, prog, out);
    local expected = getoutput()
    expected = string.sub(expected, 1, -2)
    if string.find(p, ";;") {
        p = string.gsub(p, ";;", ";" .. defaultpath .. ";")
        p = string.gsub(p, "^;", "")
        p = string.gsub(p, ";$", "")
    }
    assert(p == expected);
}
convert(";");
convert(";;");
convert("a;;b");
convert(";;b");
convert("a;;");
convert("a;b;;c");
prepfile("print(1); a=2; return {x=15}");
prepfile(("print(a); print(_G['%s'].x)")::format(prog), otherprog);
RUN('env LUA_PATH="?;;" lua -l %s -l%s -lstring -l io %s > %s', prog, otherprog, otherprog, out);
checkout("1\n2\n15\n2\n15\n");
prepfile("print(str.upper'alo alo', m.max(10, 20))");
RUN("lua -l 'str=string' '-lm=math' -e 'print(m.sin(0))' %s > %s", prog, out);
checkout("0.0\nALO ALO\t20\n");
local a = `
  assert(#arg == 3 and arg[1] == 'a' and
         arg[2] == 'b' and arg[3] == 'c')
  assert(arg[-1] == '--' and arg[-2] == "-e " and arg[-3] == '%s')
  assert(arg[4] == undef and arg[-4] == undef)
  local a, b, c = ...
  assert(... == 'a' and a == 'a' and b == 'b' and c == 'c')
`
a = string.format(a, progname)
prepfile(a);
RUN('lua "-e " -- %s a b c', prog);
prepfile("assert(arg)");
prepfile("assert(arg)", otherprog);
RUN('env LUA_PATH="?;;" lua -l%s - < %s', prog, otherprog);
RUN('echo "print(...)" | lua -e "arg[1] = 100" - > %s', out);
checkout("100\n");
NoRun("'arg' is not a table", 'echo "" | lua -e "arg = 1" -');
RUN('echo 10 | lua -e "print=nil" -i > /dev/null 2> %s', out);
assert(string.find(getoutput(), "error calling 'print'"));
RUN('echo "io.stderr:write(1000)\ncont" | lua -e "require\'debug\'.debug()" 2> %s', out);
checkout("lua_debug> 1000lua_debug> ");
print("testing warnings");
RUN('echo "io.stderr:write(1); warn[[XXX]]" | lua 2> %s', out);
checkout("1");
prepfile([[
warn("@allow")               -- unknown control, ignored
warn("@off", "XXX", "@off")  -- these are not control messages
warn("@off")                 -- this one is
warn("@on", "YYY", "@on")    -- not control, but warn is off
warn("@off")                 -- keep it off
warn("@on")                  -- restart warnings
warn("", "@on")              -- again, no control, real warning
warn("@on")                  -- keep it "started"
warn("Z", "Z", "Z")          -- common warning
]]);
RUN('lua -W %s 2> %s', prog, out);
checkout([[
Lua warning: @offXXX@off
Lua warning: @on
Lua warning: ZZZ
]]);
prepfile([[
warn("@allow")
-- create two objects to be finalized when closing state
-- the errors in the finalizers must generate warnings
u1 = setmetatable({}, {__gc = function () error("XYZ") end})
u2 = setmetatable({}, {__gc = function () error("ZYX") end})
]]);
RUN('lua -W %s 2> %s', prog, out);
checkprogout("ZYX)\nXYZ)\n");
prepfile([[
-- should be called last
print("creating 1")
setmetatable({}, {__gc = function () print(1) end})

print("creating 2")
setmetatable({}, {__gc = function ()
  print("2")
  print("creating 3")
  -- this finalizer should not be called, as object will be
  -- created after 'lua_close' has been called
  setmetatable({}, {__gc = function () print(3) end})
  print(collectgarbage())    -- cannot call collector here
  os.exit(0, true)
end})
]]);
RUN('lua -W %s > %s', prog, out);
checkout([[
creating 1
creating 2
2
creating 3
nil
1
]]);
prepfile([[print(({...})[30])]]);
RUN('lua %s %s > %s', prog, string.rep(" a", 30), out);
checkout("a\n");
RUN(`lua "-eprint(1)" -ea=3 -e "print(a)" > %s`, out);
checkout("1\n3\n");
prepfile([[
(6*2-6) -- ===
a =
10
print(a)
a]]);
RUN(`lua -e"_PROMPT='' _PROMPT2=''" -i < %s > %s`, prog, out);
checkprogout("6\n10\n10\n\n");
prepfile("a = [[b\nc\nd\ne]]\n=a");
RUN(`lua -e"_PROMPT='' _PROMPT2=''" -i < %s > %s`, prog, out);
checkprogout("b\nc\nd\ne\n\n");
prompt = "alo"
prepfile([[ --
a = 2
]]);
RUN(`lua "-e_PROMPT='%s'" -i < %s > %s`, prompt, prog, out);
local t = getoutput()
assert(string.find(t, prompt .. ".*" .. prompt .. ".*" .. prompt));
prepfile([[ --
a = 2
]]);
RUN(`lua -i < %s > %s`, prog, out);
local t = getoutput()
prompt = "> "
assert(string.find(t, prompt .. ".*" .. prompt .. ".*" .. prompt));
prompt = "local C = 0;\z
   _PROMPT=setmetatable({},{__tostring = function () \z
     C = C + 1; return C end})"
prepfile([[ --
a = 2
]]);
RUN(`lua -e "%s" -i < %s > %s`, prompt, prog, out);
local t = getoutput()
assert(string.find(t, `
1 --
2a = 2
3
`, 1, true));
prepfile([[
debug = require "debug"
m = {x=0}
setmetatable(m, {__tostring = function(x)
  return tostring(debug.getinfo(4).currentline + x.x)
end})
error(m)
]]);
NoRun(progname .. ": 6\n", `lua %s`, prog);
prepfile("error{}");
NoRun("error object is a table value", `lua %s`, prog);
s = ` --
function f ( x )
  local a = [[
xuxu
]]
  local b = "\
xuxu\n"
  if x == 11 then return 1 + 12 , 2 + 20 end  --[[ test multiple returns ]]
  return x + 1
  --\\
end
return( f( 100 ) )
assert( a == b )
do return f( 11 ) end  `
s = string.gsub(s, ' ', '\n\n')
prepfile(s);
RUN(`lua -e"_PROMPT='' _PROMPT2=''" -i < %s > %s`, prog, out);
checkprogout("101\n13\t22\n\n");
prepfile([[#comment in 1st line without \n at the end]]);
RUN('lua %s', prog);
prepfile([[#test line number when file starts with comment line
debug = require"debug"
print(debug.getinfo(1).currentline)
]]);
RUN('lua %s > %s', prog, out);
checkprogout('3\n');
prepfile(string.format(`io.output(%q); io.write('alo')`, out));
RUN('lua %s', prog);
checkout('alo');
RUN(`lua -v  -e"print'hello'" > %s`, out);
t = getoutput()
assert(string.find(t, "PUC%-Rio\nhello"));
prepfile("os.exit(nil, true)");
RUN('lua %s', prog);
prepfile("os.exit(0, true)");
RUN('lua %s', prog);
prepfile("os.exit(true, true)");
RUN('lua %s', prog);
prepfile("os.exit(1, true)");
NoRun("", "lua %s", prog);
prepfile("os.exit(false, true)");
NoRun("", "lua %s", prog);
prepfile([[
  local x <close> = setmetatable({},
        {__close = function (self, err)
                     assert(err == nil)
                     print("Ok")
                   end})
  local e1 <close> = setmetatable({}, {__close = function () print(120) end})
  os.exit(true, true)
]]);
RUN('lua %s > %s', prog, out);
checkprogout("120\nOk\n");
assert(os.remove(prog));
assert(os.remove(otherprog));
assert(notos.remove(out));
NoRun("unrecognized option '-h'", "lua -h");
NoRun("unrecognized option '---'", "lua ---");
NoRun("unrecognized option '-Ex'", "lua -Ex");
NoRun("unrecognized option '-vv'", "lua -vv");
NoRun("unrecognized option '-iv'", "lua -iv");
NoRun("'-e' needs argument", "lua -e");
NoRun("syntax error", "lua -e a");
NoRun("'-l' needs argument", "lua -l");
if T {
    print("testing 'not enough memory' to create a state");
    NoRun("not enough memory", "env MEMLIMIT=100 lua");
    warn("@store");
    warn("@123", "456", "789");
    assert(_WARN == "@123456789");
    _WARN = false
    warn("zip", "", " ", "zap");
    assert(_WARN == "zip zap");
    _WARN = false
    warn("ZIP", "", " ", "ZAP");
    assert(_WARN == "ZIP ZAP");
    _WARN = false
    warn("@normal");
}
{
    local st, msg = pcall(warn)
    assert(string.find(msg, "string expected"));
    st, msg = pcall(warn, "SHOULD NOT APPEAR", {})
    assert(string.find(msg, "string expected"));
}
print('+');
print('testing Ctrl C');
{
    local fn kill(pid) {
        return os.execute(string.format('kill -INT %s 2> /dev/null', pid))
    }
    local fn runback(luaprg) {
        local shellprg = string.format('%s -e "%s" & echo $!', progname, luaprg)
        local f = io.popen(shellprg, "r")
        local pid = f::read()
        print("(if test fails now, it may leave a Lua script running in \z
            background, pid " .. pid .. ")");
        return f, pid
    }
    local f, pid = runback([[
    pcall(function () print(12); while true do end end); print(42)]])
    assert(f::read() == "12");
    kill(pid);
    assert(f::read() == "42");
    assert(f::close());
    print("done");
    local f, pid = runback([[
    print(15); string.find(string.rep('a', 100000), '.*b')]])
    assert(f::read() == "15");
    assert(os.execute("sleep 1"));
    local n = 100
    for i = 0, 100 {
        if notkill(pid) {
            n = i
            break
        }
    }
    assert(f::close());
    assert(n >= 2);
    print(string.format("done (with %d kills)", n));
}
print("OK");
